File: web/tests/backoff.test.js
/**
 * Headless tests for web/js/backoff.js (WP 4a.2 DoD).
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { computeBackoffDelay, createBackoffState } from "../js/backoff.js";

const NO_JITTER = { jitterRatio: 0 };

test("growth: doubles per attempt with jitter disabled", () => {
  const opts = { baseMs: 100, maxMs: 1_000_000, ...NO_JITTER };
  assert.equal(computeBackoffDelay(0, opts), 100);
  assert.equal(computeBackoffDelay(1, opts), 200);
  assert.equal(computeBackoffDelay(2, opts), 400);
  assert.equal(computeBackoffDelay(3, opts), 800);
});

test("cap: never exceeds maxMs however large the attempt", () => {
  const opts = { baseMs: 100, maxMs: 1000, ...NO_JITTER };
  assert.equal(computeBackoffDelay(10, opts), 1000);
  assert.equal(computeBackoffDelay(60, opts), 1000);
  // Pathologically large attempt counts must not throw or produce NaN/Infinity
  // (2**attempt would overflow to Infinity well before this).
  assert.equal(computeBackoffDelay(500, opts), 1000);
});

test("cap holds even under jitter's upward swing", () => {
  const opts = { baseMs: 100, maxMs: 1000, jitterRatio: 0.9, random: () => 1 };
  const delay = computeBackoffDelay(10, opts); // exponential already >> maxMs
  assert.ok(delay <= 1000, `expected <= 1000, got ${delay}`);
});

test("jitter bounds: swings within +/- jitterRatio of the capped value", () => {
  // attempt=1 with base=1000 -> exponential=2000, well under maxMs, so the
  // cap does not interfere and the full jitter span is observable.
  const opts = { baseMs: 1000, maxMs: 100_000, jitterRatio: 0.5 };
  const capped = 2000;
  const span = capped * 0.5;

  const low = computeBackoffDelay(1, { ...opts, random: () => 0 });
  const mid = computeBackoffDelay(1, { ...opts, random: () => 0.5 });
  const high = computeBackoffDelay(1, { ...opts, random: () => 1 });

  assert.equal(low, capped - span);
  assert.equal(mid, capped);
  assert.equal(high, capped + span);
});

test("jitter never produces a negative delay (floor is load-bearing)", () => {
  // jitterRatio > 1 so the UNFLOORED math actually goes negative:
  // capped=100, span=150, swing=-150 -> jittered=-50. A mutant that deletes
  // the `Math.max(0, ...)` floor in backoff.js would return -50 here, not
  // merely "still happens to be >= 0" (jitterRatio=1 previously landed
  // exactly on 0 either way, which a removed floor could not be told apart
  // from — vacuous, WP 4a.2 review should-fix #1).
  const opts = { baseMs: 100, maxMs: 100_000, jitterRatio: 1.5, random: () => 0 };
  const delay = computeBackoffDelay(0, opts);
  assert.equal(delay, 0, `expected the floor to clamp to exactly 0, got ${delay}`);
});

test("rejects a negative or non-integer attempt", () => {
  assert.throws(() => computeBackoffDelay(-1), RangeError);
  assert.throws(() => computeBackoffDelay(1.5), RangeError);
  assert.throws(() => computeBackoffDelay(NaN), RangeError);
});

test("createBackoffState: next() advances the attempt counter", () => {
  const opts = { baseMs: 100, maxMs: 100_000, jitterRatio: 0 };
  const state = createBackoffState(opts);
  assert.equal(state.attempt, 0);
  assert.equal(state.next(), 100); // attempt 0
  assert.equal(state.attempt, 1);
  assert.equal(state.next(), 200); // attempt 1
  assert.equal(state.next(), 400); // attempt 2
  assert.equal(state.attempt, 3);
});

test("createBackoffState: reset() on success returns to the fast schedule", () => {
  const opts = { baseMs: 100, maxMs: 100_000, jitterRatio: 0 };
  const state = createBackoffState(opts);
  state.next();
  state.next();
  state.next();
  assert.equal(state.attempt, 3);
  state.reset();
  assert.equal(state.attempt, 0);
  assert.equal(state.next(), 100); // back to attempt-0 delay
});

File: app/app/src/test/java/dev/steamvault/app/polling/BackoffTest.kt
package dev.steamvault.app.polling

import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Growth/cap/jitter math for [Backoff], both directions of every boundary
 * (docs/LEARNINGS.md "Testing discipline") — a direct port of
 * `web/tests/backoff.test.js`'s coverage for the same defaults
 * (1 s base, 30 s cap, 20% jitter ratio), since [Backoff] is a
 * line-for-line port of `web/js/backoff.js`.
 */
class BackoffTest {

    private val noJitter = BackoffOptions(jitterRatio = 0.0)

    @Test
    fun `attempt 0 with no jitter is exactly baseMs`() {
        assertEquals(1000L, Backoff.computeDelayMs(0, noJitter))
    }

    @Test
    fun `growth is exponential -- baseMs times 2 to the attempt, with no jitter`() {
        assertEquals(1000L, Backoff.computeDelayMs(0, noJitter))
        assertEquals(2000L, Backoff.computeDelayMs(1, noJitter))
        assertEquals(4000L, Backoff.computeDelayMs(2, noJitter))
        assertEquals(8000L, Backoff.computeDelayMs(3, noJitter))
    }

    @Test
    fun `growth is capped at maxMs, not left to grow unbounded`() {
        val delay = Backoff.computeDelayMs(10, noJitter) // 1000 * 2^10 = 1,024,000 uncapped
        assertEquals(30000L, delay)
    }

    @Test
    fun `a pathologically large attempt count does not overflow to a garbage delay`() {
        // 2^63 would overflow a Long; the >62 guard routes this through
        // Double.POSITIVE_INFINITY before min() clamps it back to maxMs.
        val delay = Backoff.computeDelayMs(1000, noJitter)
        assertEquals(30000L, delay)
    }

    @Test
    fun `attempt must be non-negative`() {
        assertThrows(IllegalArgumentException::class.java) {
            Backoff.computeDelayMs(-1, noJitter)
        }
    }

    @Test
    fun `jitter never pushes the delay above maxMs, even with a maximal upward swing`() {
        val alwaysMax = BackoffOptions(jitterRatio = 0.2, random = { 1.0 }) // swing = +span, the upward extreme
        val delay = Backoff.computeDelayMs(10, alwaysMax) // capped value is already maxMs
        assertEquals(30000L, delay)
    }

    @Test
    fun `jitter never pushes the delay below zero, even with a maximal downward swing`() {
        val alwaysMin = BackoffOptions(baseMs = 100, jitterRatio = 1.0, random = { 0.0 }) // swing = -span = -capped
        val delay = Backoff.computeDelayMs(0, alwaysMin)
        assertTrue(delay >= 0)
        assertEquals(0L, delay)
    }

    @Test
    fun `jitter floor is load-bearing -- jitterRatio greater than 1 would go NEGATIVE without it`() {
        // WP 4b.2 Opus review should-fix: the test above (jitterRatio=1.0)
        // is VACUOUS for pinning the floor -- capped(100) - span(100) lands
        // on exactly 0 whether or not `max(0.0, ...)` runs at all, so
        // deleting that floor from Backoff.kt still passes it. This test
        // mirrors web/tests/backoff.test.js's "jitter never produces a
        // negative delay (floor is load-bearing)" case exactly
        // (baseMs=100, maxMs=100_000, jitterRatio=1.5, random=0, attempt=0):
        // capped=100, span=150, swing=-150, unfloored jittered=-50. Only
        // the floor turns that -50 into the 0 asserted here -- deleting
        // `max(0.0, ...)` in Backoff.kt makes this test fail.
        val options = BackoffOptions(baseMs = 100, maxMs = 100_000, jitterRatio = 1.5, random = { 0.0 })
        val delay = Backoff.computeDelayMs(0, options)
        assertEquals(0L, delay)
    }

    @Test
    fun `zero jitterRatio returns the capped value exactly, no randomness involved`() {
        val options = BackoffOptions(jitterRatio = 0.0, random = { error("must not be called when jitterRatio is 0") })
        assertEquals(4000L, Backoff.computeDelayMs(2, options))
    }

    @Test
    fun `BackoffState next increments the attempt counter and reset zeroes it`() {
        val state = BackoffState(noJitter)

        assertEquals(1000L, state.next())
        assertEquals(2000L, state.next())
        assertEquals(4000L, state.next())
        assertEquals(3, state.attempt)

        state.reset()

        assertEquals(0, state.attempt)
        assertEquals(1000L, state.next())
    }
}
